File: crates/fey_math/src/ray.rs
use crate::{Num, Vec2, impl_approx, impl_bytemuck, impl_casts};
use serde::{Deserialize, Serialize};

pub type RayF = Ray<f32>;

/// A ray with an origin and direction.
#[repr(C)]
#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
pub struct Ray<T> {
    pub origin: Vec2<T>,
    pub direction: Vec2<T>,
}

impl_bytemuck!(Ray);

impl_approx!(
    NAME = Ray
    FIELDS = (origin, direction)
);

impl_casts!(
    NAME = Ray
    FIELDS = (origin, direction)
);

/// Create a [`Ray`].
#[inline]
pub const fn ray<T>(origin: Vec2<T>, direction: Vec2<T>) -> Ray<T> {
    Ray { origin, direction }
}

impl<T> Ray<T> {
    /// Create a new ray.
    #[inline]
    pub const fn new(origin: Vec2<T>, direction: Vec2<T>) -> Self {
        ray(origin, direction)
    }
}

impl<T: Num> Ray<T> {
    /// Get a point at the distance `dist` along the ray.
    #[inline]
    pub fn point(&self, dist: T) -> Vec2<T> {
        self.origin + self.direction * dist
    }
}

File: crates/fey_math/lua/Ray.lua
---@meta

---A ray with an origin and direction.
---@class (exact) Ray: RayMethods
---@field origin Vec2
---@field direction Vec2

---@class RayClass : RayMethods
---@overload fun(origin: Vec2, direction: Vec2): Ray
local module = {}

---@class RayMethods
local methods = {}

---Create a new ray.
---@param origin Vec2
---@param direction Vec2
---@return Ray
---@nodiscard
function module.new(origin, direction) end

---Returns a temporary copy of this value.
---@param self Ray
---@return Ray
---@nodiscard
function methods.clone(self) end

---Boxes the value if not already boxed.
---@param self Ray
---@return Ray
---@nodiscard
function methods.box(self) end

---Boxes a copy of this value.
---@param self Ray
---@return Ray
---@nodiscard
function methods.box_clone(self) end

---Returns `true` if the two rays are approximately equal.
---@param self Ray
---@param other Ray
---@return boolean
---@nodiscard
function methods.approx(self, other) end

---Get a point at the `distance` along the ray.
---@param self Ray
---@param distance number
---@return Vec2
---@nodiscard
function methods.point(self, distance) end

return module